docs/source/about_mapstyle_vs_iterable.mdx

```diff
@@ -0,0 +1,220 @@
+for example in my_iterable_dataset:  
+    print(example)
+    break
+```
+
+## Exact and fast approximate shuffling
+
+When you shuffle a [`Dataset`] using [`Dataset.shuffle`], you apply an exact shuffling of the dataset.
+It works by taking a list of indices `[0, 1, 2, ... len(my_dataset) - 1]` and shuffling this list.
+Then, accessing `my_dataset[0]` accesses the row and the index defined by the first element of the indices mapping that has been shuffled:
+```python
+my_dataset = my_dataset.shuffle(seed=42)
+print(my_dataset[0])
+```
+
+Since we don't have random access to the rows in the case of iterable datasets, we can't use a shuffled list of indices and access a row at an arbitrary position.
+This prevents the use of exact shuffling.
+Instead, a fast approximate shuffling is used in [`IterableDataset.shuffle`].
+It uses a shuffle buffer to sample random examples iteratively from the dataset.
+Since the dataset is still read iteratively, it provides excellent speed performance:
+```python
+my_iterable_dataset = my_iterable_dataset.shuffle(seed=42, buffer_size=100)
+for example in my_iterable_dataset:
+    print(example)
+    break
+```
+
+Since using a shuffle buffer is not enough to provide a satisfactory shuffling for machine learning model training, [`IterableDataset.shuffle`] also shuffles the dataset shards if your dataset is made of multiple files or sources:
+
+```python
+# Stream from the internet
+my_iterable_dataset = load_dataset("c4", "en", split="train", streaming=True)
+my_iterable_dataset.n_shards  # 1024
+
+# Stream from local files
+data_files = {"train": [f"path/to/data_{i}.csv" for i in range(1024)]}
+my_iterable_dataset = load_dataset("csv", data_files=data_files, split="train", streaming=True)
+my_iterable_dataset.n_shards  # 1024
+
+# From a generator function
+def my_generator(n, sources):
+    for source in sources:
+        for example_id_for_current_source in range(n):
+            yield {"example_id": f"{source}_{example_id_for_current_source}"}
+
+gen_kwargs = {"n": 10, "sources": [f"path/to/data_{i}" for i in range(1024)]}
+my_iterable_dataset = IterableDataset.from_generator(my_generator, gen_kwargs=gen_kwargs)
+my_iterable_dataset.n_shards  # 1024
+```
+
+## Speed differences
+
+Regular [`Dataset`] objects are based on Arrow which provides fast random access to the rows.
+Thanks to memory mapping and the fact that Arrow is an in-memory format, reading data from disk doesn't do expensive system calls and deserialization.
+
+It also provides even faster data loading when iterating using a `for` loop by iterating on contiguous Arrow record batches.
+
+However as soon as your [`Dataset`] has an indices mapping (via [`Dataset.shuffle`] for example), the speed can become 10x slower.
+This is because there is an extra step to get the row index to read using the indices mapping, and most importantly because you no longer read contiguous chunks of data.
```

Comment:
```suggestion
This is because there is an extra step to get the row index to read using the indices mapping, and most importantly, you aren't reading contiguous chunks of data anymore.
```